### PR TITLE
[dynamo][guards-cpp-refactor] TupleIteratorGetItemAccessor

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120469
* #120416
* #120359
* #120344
* #120342
* #120096
* #120093
* #120123
* #120119
* __->__ #120091
* #120089
* #120068
* #120067
* #120065
* #120064
* #120062
* #120061
* #120060
* #119833
* #119827
* #119822

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/csrc/dynamo/guards.cpp [hunk 1]
```diff
@@ -1687,6 +1710,55 @@ class TypeGuardAccessor : public GuardAccessor {
...
+      RootGuardManager* root,
+      py::object index,
+      py::handle example_value)
+      : GuardAccessor(root, index, example_value),
+        _index(py::cast<Py_ssize_t>(index)) {}
+
+  // NB: Intentional duplication between check_nopybind and
+  // check_verbose_nopybind.
+  bool check_nopybind(PyObject* obj) override { // borrowed ref
+    _PyTupleIterObject* it = (_PyTupleIterObject*)obj;
```
Comment (jansel) [inline torch/csrc/dynamo/guards.cpp:1728]: Do we need assert on the type of obj first?

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/csrc/dynamo/guards.cpp (hunk 1 above)
Comment (anijain2305) [reply torch/csrc/dynamo/guards.cpp:1733]: We don't really need to. If we are going through this Accessor, we would have already put a leaf (or trunk guard) on the tuple iterator itself. This guard would check for TYPE_MATCH.

image

Let me know if you still want it.

Comment (jansel) [reply torch/csrc/dynamo/guards.cpp:1733]: Yeah, let's merge those into one guard.  Should be slightly faster.

Comment (anijain2305) [reply torch/csrc/dynamo/guards.cpp:1733]: If we do want to combine the two guards, I believe we should do it in TUPLE_ITERATOR_LEN guard. 

This PR is for `TupleIteratorGetItemAcccessor`. The structure will be something like

- GuardManager for tuple_iterator
     - Leaf_guard - TYPE_MATCH
     - Leaf_guard - TUPLE_ITERATOR_LEN
     - GuardManager for an item on tuple-iterator via TupleIteratorGetItemAccessor
            - Leaf guard - (like a constant_match or something)


So, I don't think we should add a type_check in `TupleIteratorGetItemAccessor`. We will do one more redundant TYPE_MATCH.

Comment (anijain2305) [reply torch/csrc/dynamo/guards.cpp:1733]: I will merge the TYPE_MATCH guard with `TUPLE_ITERATOR_LEN` guard in this PR - https://github.com/pytorch/pytorch/pull/120119

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**